clang/test/Sema/struct-cast.c

```diff
@@ -1,4 +1,5 @@
 // RUN: %clang_cc1 -fsyntax-only %s -verify
+// RUN: %clang_cc1 -fsyntax-only %s -fexperimental-new-constant-interpreter -verify
```

Comment:
I'm really confused what this test has to do with the changes above. The changes relate to three-way comparison operators but this test is a C test file and has no comparison operators involved. Why would this test hit that assertion?

Reply:
The AST we trip up on is
```
BinaryOperator 0x52100006ed08 'struct S' '='
|-DeclRefExpr 0x52100006eba0 'struct S' lvalue Var 0x52100006e998 'tmp' 'struct S'
`-CStyleCastExpr 0x52100006ecd8 'struct S' <NoOp>
  `-CallExpr 0x52100006ec88 'struct S'
    `-ImplicitCastExpr 0x52100006ec68 'const struct S (*)(void)' <FunctionToPointerDecay>
      `-DeclRefExpr 0x52100006ebc8 'const struct S (void)' Function 0x52100006e870 'foo' 'const struct S (void)'
```

IIUC this should be rejected since the assignment doesn't work, but we're asserting that we found `CmpInfo`. We could of course also add more assertions that we're even in C++ mode when looking for the spaceship operator stuff.

Reply:
Why should this assignment not work? (The code looks valid to me). Oh, wait, you mean this assignment shouldn't be a constant expression? (If so, I agree.)

That said, I see now why we're hitting the assertion on an assignment operator in C; the call to `lookupInfoForType()` will return `nullptr` in that case and we need to handle that.

I don't think we want to look for the spaceship operator stuff outside of C++ mode, but that can be done in a follow-up as this is fixing a failing assertion.

Reply:
Yes, what I mean is, the constant interpreter should return `false` for this assignment in C.

Reply:
Excellent, thank you!